### PR TITLE
feat: Add support for CM provider, autocertificates and add tests

### PR DESCRIPTION
**Pull request by adeatcu-ionos**
## Checklist

- [ ] PR name added as appropriate (e.g. `feat:`/`fix:`/`doc:`/`test:`/`refactor:`)
- [ ] Tests added or updated
- [ ] Documentation updated
- [ ] Changelog updated and version incremented (label: upcoming release)
- [ ] Github Issue linked if any
- [ ] Jira task updated

### CODE REVIEW

File: docs/data-sources/certificate_manager_auto_certificate.md
Comment (avirtopeanu-ionos) [inline docs/data-sources/certificate_manager_auto_certificate.md:1]: These docs need to be added to `gitbook_docs/summary.md`

File: ionoscloud/resource_certificate_manager_auto_certificate_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+				Config: CMAutoCertificateConfig,
+				Check: resource.ComposeTestCheckFunc(
+					CMAutoCertificateExistenceCheck(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, &autoCertificate),
+					resource.TestCheckResourceAttr(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, CMProviderLocationAttr, CMProviderLocationVal),
+					resource.TestCheckResourceAttr(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, CMAutoCertCommonNameAttr, CMAutoCertCommonNameVal),
+					resource.TestCheckResourceAttr(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, CMAutoCertKeyAlgorithmAttr, CMAutoCertKeyAlgorithmVal),
+					resource.TestCheckResourceAttr(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, CMAutoCertNameAttr, CMAutoCertNameVal),
+					resource.TestCheckResourceAttrSet(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, "provider_id"),
+					resource.TestCheckResourceAttrSet(constant.AutoCertificateResource+"."+constant.TestCMAutoCertificateName, "last_issued_certificate_id"),
+					// TODO -- Check that subject_alternative_names contains the proper list
```
Comment (avirtopeanu-ionos) [inline ionoscloud/resource_certificate_manager_auto_certificate_test.go:36]: is this TODO also meant to be merged to master?

File: utils/constant/constants.go [hunk 2]
```diff
@@ -347,8 +347,12 @@ const (
 const ServersDataSource = "ionoscloud_servers"
 
 const (
-	CertificateResource = "ionoscloud_certificate"
-	TestCertName        = "certTest"
+	CertificateResource             = "ionoscloud_certificate"
+	AutoCertificateProviderResource = "ionoscloud_auto_certificate_provider"
+	AutoCertificateResource         = "ionoscloud_auto_certificate"
+	TestCertName                    = "certTest"
```
Comment (avirtopeanu-ionos) [inline utils/constant/constants.go:353]: these seem to be used only in a single test file, is there really any reason for having them in the constants pkg

this pkg is imported many times in the code base

File: ionoscloud/resource_certificate_manager_auto_certificate.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+}
+
+func autoCertificateImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	client := meta.(services.SdkBundle).CertManagerClient
+	parts := strings.Split(d.Id(), ":")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid import ID: %v, expected ID in the format: '<location>:<auto_certificate_ID>", d.Id())
+	}
+	location := parts[0]
+	if !slices.Contains(constant.MariaDBClusterLocations, location) {
```
Comment (cbeti-ionos) [inline ionoscloud/resource_certificate_manager_auto_certificate.go:173]: use the new location constant

File: ionoscloud/data_source_certificate_manager_auto_certificate.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
+package ionoscloud
+
+import (
+	"context"
```
Comment (cbeti-ionos) [inline ionoscloud/data_source_certificate_manager_auto_certificate.go:4]: sort the imports on the new files

File: ionoscloud/data_source_certificate_manager_auto_certificate.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	certSDK "github.com/ionos-cloud/sdk-go-cert-manager"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/services"
+	certService "github.com/ionos-cloud/terraform-provider-ionoscloud/v6/services/cert"
+	"github.com/ionos-cloud/terraform-provider-ionoscloud/v6/utils/constant"
+	"strings"
+)
```
Comment (cbeti-ionos) [inline ionoscloud/data_source_certificate_manager_auto_certificate.go:14]: sort the imports in the new files

**Review by avirtopeanu-ionos — DISMISSED** [review]

File: utils/constant/constants.go (hunk 2 above)
Comment (adeatcu-ionos) [reply utils/constant/constants.go:359]: @avirtopeanu-ionos it is also used in `import_certificate_manager_certificate_test.go`

File: docs/data-sources/certificate_manager_auto_certificate.md
Comment (adeatcu-ionos) [reply docs/data-sources/certificate_manager_auto_certificate.md:1]: @avirtopeanu-ionos added in 151f8a269d16c58497218a43943de98ba9eb83f9

File: ionoscloud/resource_certificate_manager_auto_certificate_test.go (hunk 1 above)
Comment (adeatcu-ionos) [reply ionoscloud/resource_certificate_manager_auto_certificate_test.go]: @avirtopeanu-ionos no, thanks for catching this. I added this `TODO` but forgot to complete it. It is now completed in 151f8a269d16c58497218a43943de98ba9eb83f9.

File: ionoscloud/resource_certificate_manager_auto_certificate.go (hunk 3 above)
Comment (adeatcu-ionos) [reply ionoscloud/resource_certificate_manager_auto_certificate.go]: @cbeti-ionos done

File: ionoscloud/data_source_certificate_manager_auto_certificate.go (hunk 5 above)
Comment (adeatcu-ionos) [reply ionoscloud/data_source_certificate_manager_auto_certificate.go]: @cbeti-ionos done

File: ionoscloud/data_source_certificate_manager_auto_certificate.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+			for _, autoCertificateItem := range *autoCertificates.Items {
+				if autoCertificateItem.Properties != nil && autoCertificateItem.Properties.Name != nil && strings.EqualFold(*autoCertificateItem.Properties.Name, name.(string)) {
+					results = append(results, autoCertificateItem)
+				}
+			}
+		}
+
+		if len(results) == 0 {
+			return diag.FromErr(fmt.Errorf("no auto-certificate found with the specified name: %v", name))
+		} else if len(results) > 1 {
```
Comment (digna-ionos) [inline ionoscloud/data_source_certificate_manager_auto_certificate.go:111]: you can use just if without the else

File: ionoscloud/data_source_certificate_manager_provider.go [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	if !idOk && !nameOk {
+		return diag.FromErr(fmt.Errorf("please provide either the auto-certificate provider ID or name"))
+	}
+
+	var provider certSDK.ProviderRead
+	var err error
+
+	if idOk {
+		id := id.(string)
```
Comment (digna-ionos) [inline ionoscloud/data_source_certificate_manager_provider.go:85]: you already have the id

File: ionoscloud/data_source_certificate_manager_auto_certificate.go (hunk 6 above)
Comment (adeatcu-ionos) [reply ionoscloud/data_source_certificate_manager_auto_certificate.go:110]: @digna-ionos please leave a suggestion with how the code should look like, it is not clear for me from your comment. I'm interested in the case in which the `len(results) is 1`, the error cases are `== 0` and `> 1`, not sure how can I do this check without an if/else, so please leave the exact code so I can understand properly. Thanks!

Comment (digna-ionos) [reply ionoscloud/data_source_certificate_manager_auto_certificate.go:110]: `
if len(results) == 0 {
  return ...
}
if len(results) > 1 {
  return ...
}
`

File: ionoscloud/resource_certificate_manager_provider.go [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+					},
+				},
+			},
+		},
+		CustomizeDiff: checkProviderImmutableFields,
+		Timeouts:      &resourceDefaultTimeouts,
+	}
+}
+
+func checkProviderImmutableFields(_ context.Context, diff *schema.ResourceDiff, _ interface{}) error {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_certificate_manager_provider.go:76]: let's make these forcenew as is idiomatic. I know we have some older resources that throw errors, but new ones should force replace

File: ionoscloud/resource_certificate_manager_auto_certificate.go [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+				Computed:    true,
+				Description: "The ID of the last certificate that was issued",
+			},
+		},
+		CustomizeDiff: checkAutoCertificateImmutableFields,
+		Timeouts:      &resourceDefaultTimeouts,
+	}
+}
+
+func checkAutoCertificateImmutableFields(_ context.Context, diff *schema.ResourceDiff, _ interface{}) error {
```
Comment (cristiGuranIonos) [inline ionoscloud/resource_certificate_manager_auto_certificate.go:75]: same, make them forcenew

File: ionoscloud/data_source_certificate_manager_auto_certificate.go (hunk 6 above)
Comment (adeatcu-ionos) [reply ionoscloud/data_source_certificate_manager_auto_certificate.go:110]: @digna-ionos thanks, I'll make the change.

File: services/cert/provider.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+	location := d.Get("location").(string)
+	provider, apiResponse, err := c.GetProvider(ctx, providerID, location)
+	if err != nil {
+		if apiResponse.HttpNotFound() {
+			return true, nil
+		}
+		return false, fmt.Errorf("error while checking deletion status for certificate manager provider with ID: %v, error: %w", d.Id(), err)
+	}
+	if provider.Metadata != nil && provider.Metadata.State != nil {
+		if utils.IsStateFailed(*provider.Metadata.State) {
```
Comment (cbeti-ionos) [inline services/cert/provider.go:99]: can merge these into a single `if`

File: ionoscloud/resource_certificate_manager_provider.go [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+}
+
+func providerImport(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	client := meta.(services.SdkBundle).CertManagerClient
+	parts := strings.Split(d.Id(), ":")
+	if len(parts) != 2 {
+		return nil, fmt.Errorf("invalid import ID: %v, expected ID in the format: '<location>:<provider_id>'", d.Id())
+	}
+	location := parts[0]
+	// TODO -- Modify the name of this constant
```
Comment (cbeti-ionos) [inline ionoscloud/resource_certificate_manager_provider.go:177]: resolve TODO or remove it if no longer needed

File: services/cert/auto_certificate.go [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+	autoCertificateID := d.Id()
+	location := d.Get("location").(string)
+	autoCertificate, apiResponse, err := c.GetAutoCertificate(ctx, autoCertificateID, location)
+	if err != nil {
+		if apiResponse.HttpNotFound() {
+			return true, nil
+		}
+		return false, fmt.Errorf("error while checking deletion status for auto-certificate with ID: %v, error: %w", d.Id(), err)
+	}
+	if utils.IsStateFailed(*autoCertificate.Metadata.State) {
```
Comment (cbeti-ionos) [inline services/cert/auto_certificate.go:86]: is a nilcheck required here before the derefernce?

File: ionoscloud/data_source_certificate_manager_auto_certificate.go (hunk 6 above)
Comment (adeatcu-ionos) [reply ionoscloud/data_source_certificate_manager_auto_certificate.go]: @digna-ionos done in ec747230649ad96067a92e1d3b11d469d3eda2f1

File: ionoscloud/resource_certificate_manager_provider.go (hunk 8 above)
Comment (adeatcu-ionos) [reply ionoscloud/resource_certificate_manager_provider.go]: @cristiGuranIonos done in ec747230649ad96067a92e1d3b11d469d3eda2f1

File: ionoscloud/resource_certificate_manager_auto_certificate.go (hunk 9 above)
Comment (adeatcu-ionos) [reply ionoscloud/resource_certificate_manager_auto_certificate.go]: @cristiGuranIonos done in ec747230649ad96067a92e1d3b11d469d3eda2f1

File: services/cert/provider.go (hunk 10 above)
Comment (adeatcu-ionos) [reply services/cert/provider.go]: @cbeti-ionos done in ec747230649ad96067a92e1d3b11d469d3eda2f1

File: ionoscloud/resource_certificate_manager_provider.go (hunk 11 above)
Comment (adeatcu-ionos) [reply ionoscloud/resource_certificate_manager_provider.go]: @cbeti-ionos done in ec747230649ad96067a92e1d3b11d469d3eda2f1

File: services/cert/auto_certificate.go (hunk 12 above)
Comment (adeatcu-ionos) [reply services/cert/auto_certificate.go]: @cbeti-ionos done in ec747230649ad96067a92e1d3b11d469d3eda2f1

**Pull request merged**